File: backend/app/routers/coin_packages.py
"""Coin package CRUD endpoints (admin panel)."""

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user, require_admin
from app.models.user import User
from app.schemas.coin_package import (
    CoinPackageCreate,
    CoinPackageListResponse,
    CoinPackageResponse,
    CoinPackageUpdate,
)
from app.services import coin_service

router = APIRouter(prefix="/api/coin-packages", tags=["coin-packages"])


@router.get("", response_model=CoinPackageListResponse)
async def list_packages(
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """List all coin packages."""
    packages = await coin_service.list_packages(db)
    return CoinPackageListResponse(items=packages, total=len(packages))


@router.post("", response_model=CoinPackageResponse, status_code=status.HTTP_201_CREATED)
async def create_package(
    request: CoinPackageCreate,
    admin: Annotated[User, Depends(require_admin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Create a new coin package. Admin only."""
    return await coin_service.create_package(
        db,
        name=request.name,
        description=request.description,
        coin_amount=request.coin_amount,
        price_sar=request.price_sar,
        sort_order=request.sort_order,
        created_by=admin.id,
    )


@router.patch("/{package_id}", response_model=CoinPackageResponse)
async def update_package(
    package_id: uuid.UUID,
    request: CoinPackageUpdate,
    _admin: Annotated[User, Depends(require_admin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Update a coin package. Admin only."""
    update_data = request.model_dump(exclude_unset=True)
    package = await coin_service.update_package(db, package_id, **update_data)
    if package is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Coin package not found"
        )
    return package


@router.delete("/{package_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_package(
    package_id: uuid.UUID,
    _admin: Annotated[User, Depends(require_admin)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Soft-delete a coin package (set is_active=false). Admin only."""
    found = await coin_service.deactivate_package(db, package_id)
    if not found:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Coin package not found"
        )

File: frontend/src/stores/coinPackages.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'
import axios from 'axios'
import api from '../lib/api'
import type {
  CoinPackage,
  CoinPackageListResponse,
  CoinPackageCreate,
  CoinPackageUpdate,
} from '../types/subscriber'

function extractError(e: unknown, fallback: string): string {
  if (axios.isAxiosError(e)) return e.response?.data?.detail ?? fallback
  return fallback
}

export const useCoinPackageStore = defineStore('coinPackages', () => {
  // State
  const packages = ref<CoinPackage[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const total = ref(0)

  // Actions
  async function fetchPackages() {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<CoinPackageListResponse>('/api/coin-packages')
      packages.value = data.items
      total.value = data.total
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to fetch coin packages')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function createPackage(payload: CoinPackageCreate) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.post<CoinPackage>('/api/coin-packages', payload)
      packages.value.push(data)
      total.value++
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to create coin package')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updatePackage(id: string, payload: CoinPackageUpdate) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.patch<CoinPackage>(`/api/coin-packages/${id}`, payload)
      const index = packages.value.findIndex(p => p.id === id)
      if (index !== -1) packages.value[index] = data
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to update coin package')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deletePackage(id: string) {
    loading.value = true
    error.value = null
    try {
      await api.delete(`/api/coin-packages/${id}`)
      const index = packages.value.findIndex(p => p.id === id)
      if (index !== -1) packages.value[index].is_active = false
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to deactivate coin package')
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    // State
    packages,
    loading,
    error,
    total,
    // Actions
    fetchPackages,
    createPackage,
    updatePackage,
    deletePackage,
  }
})
